lib/roby/relations/incremental_transitive_closure.rb

```diff
@@ -0,0 +1,140 @@
+        # where edges and vertices can be added or removed incrementally, 
+        # while keeping track of reachability information (i.e., transitive closure).
+        class IncrementalTransitiveClosure
+            attr_reader :graph
+
+            # Initializes an empty directed graph using RGL's DirectedAdjacencyGraph
+            def initialize
+                @graph = RGL::DirectedAdjacencyGraph.new
+            end
+        
+            # Adds a new vertex to the graph if it doesn't already exist
+            #
+            # @param vertex [Object] The vertex to be added
+            def added_vertex(vertex)
+                return if @graph.has_vertex?(vertex)
+
+                @graph.add_vertex(vertex)
+            end
+            
+            # Adds an edge from 'source' to 'target' to the graph
+            # It also ensures that transitive reachability is updated by adding edges
+            # from the 'source's parents vertices to 'target'. Adding edges
+            # from the from 'source' to 'target's children vertices.
+            # As well as adding edges from 'source's parent vertices to 'target's children
+            #
+            # @param source [Object] The source vertex
+            # @param target [Object] The target vertex
+            def added_edge(source, target)
+                return if @graph.has_edge?(source, target)
+
+                @graph.add_edge(source, target)
+
+                parents = parent_vertices(source)
+                add_edges_vertices_to_vertex(parents, target)
+
+                children = @graph.adjacent_vertices(target)
+                add_edges_vertex_to_vertices(source, children)
+
+                add_edges_vertices_to_vertices(parents, children)
+            end
+
+            # Removes a vertex from the graph if it has no adjacent vertices 
+            # (i.e., no outgoing edges)
+            # If there are adjacent vertices, the entire graph is 
+            # reset (rebuilds the graph)
+            #
+            # @param vertex [Object] The vertex to be removed
+            def removed_vertex(vertex)
+                return unless @graph.has_vertex?(vertex)
+
+                if @graph.adjacent_vertices(vertex).empty?
+                    @graph.remove_vertex(vertex)
+                else
+                    @graph = RGL::DirectedAdjacencyGraph.new
+                end
+            end
+
+            # Removes an edge from the graph. If the target vertex has adjacent vertices,
+            # or the source vertex has parent vertices
+            # the entire graph is reset. Otherwise, it simply removes the edge.
+            #
+            # @param source [Object] The source vertex
+            # @param target [Object] The target vertex
+            def removed_edge(source, target)
+                return unless @graph.has_edge?(source, target)
+
+                if @graph.adjacent_vertices(target).empty? && parent_vertices(source).empty?
```

Comment:
Whenever you enumerate something to test if it is `empty?`, you should be asking yourself whether there is not a better way. This is not only in Ruby, it's general.

In this case, graphs have `root?` and `leaf?`